src/Data/Array/Accelerate/Data/Either.hs

```diff
@@ -89,16 +83,14 @@ isRight x = tag x == 1
 -- instead.
 --
 fromLeft :: (Elt a, Elt b) => Exp (Either a b) -> Exp a
-fromLeft x = a
-  where T3 _ a _ = asTuple x
+fromLeft (Exp e) = Exp $ SmartExp $ Prj PairIdxRight $ SmartExp $ Prj PairIdxLeft $ SmartExp $ Prj PairIdxRight e
```

Comment:
I thought the goal of this PR was to allow users to define custom sum types. I hope they don't have to write this? Or is this line just avoiding generics, to be more efficient?

Reply:
nah, this is just to define the unsafe `fromLeft` and `fromRight` functions. These existed from before this patch, I'm just updating them.

Reply:
I guess what I mean is, if defining 'your own' Either datatype (including all these 'isRight' and 'fromLeft' functions) is now possible on the user side, it would make sense to show off the `intended' way here instead of using Accelerate internals? Now that this vesion already works, that might be wasted effort, but I was surprised to see this after the paper promised a very neat interface

Reply:
ah, I understand what you mean now. I updated the functions in that module and added an example to the 'match' docs.

Reply:
`isRight (Exp e) = Exp $ SmartExp $ (SmartExp $ Prj PairIdxLeft e) `Pair` SmartExp Nil`
I assume this is the same as something like:
`isRight = match \case`
`  Right_ _ -> True`
`  Left_   _ -> False`

This looks great! I was already planning on rewriting the sudoku solver, now that I know more about Accelerate (even changing a 'run' to 'runN' should already be an improvement I think, it now repeatedly compiles the same Acc program), and I might see what happens to the readability and performance if I use some nice ADTs instead of bitwise hacking.

Reply:
Yes, exactly. I'm just cheating because I know the tag bits for `Right` and `True` happen to be the same. I'll add a comment mentioning this.

At some point I went and added pattern synonyms to the sudokus code. Apparently I left it in a half-committed state, but I put it here: https://github.com/tmcdonell/Sudokus

Reply:
Oh, that already looks much better! I'll put some more effort into that once I'm done :)